c++ training/LinkedLists/LinkedLists/DoublyLinkedList.h

```diff
@@ -0,0 +1,323 @@
+		}
+
+		if (Index == 0)
+		{
+			AddHead(Data);
+			return;
+		}
+		else
+		{
+			FNode& PreviousNode = GetNodeAt(Index - 1);
+			FNode* CurrentPtr = PreviousNode.Next;
+			PreviousNode.Next = new FNode(Data, CurrentPtr, &PreviousNode);
+
+			if (Index == Size)
+			{
+				Tail = Tail->Next;
+			}
+			else
+			{
+				CurrentPtr->Previous = PreviousNode.Next;
+			}
+
+			Size++;
+		}
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	void Remove(const int Index)
+	{
+		if (Index < 0 || Index >= Size)
+		{
+			return;
+		}
+
+		if (Index == 0)
+		{
+			FNode* NodeToDeletePtr = Head;
+			Head = NodeToDeletePtr->Next;
+
+			if (Head)
+			{
+				Head->Previous = nullptr;
+			}
+
+			delete NodeToDeletePtr;
+			NodeToDeletePtr = nullptr;
+
+			if (Size == 1)
+			{
+				Tail = nullptr;
+			}
+		}
+		else
+		{
+			FNode& PreviousNode = GetNodeAt(Index - 1);
+			FNode* NodeToDeletePtr = PreviousNode.Next;
+			FNode* NewNextPtr = NodeToDeletePtr->Next;
+
+			PreviousNode.Next = NewNextPtr;
+			
+			if (NewNextPtr)
+			{
+				NewNextPtr->Previous = &PreviousNode;
+			}
+			else
+			{
+				Tail = &PreviousNode;
+			}
+
+			delete NodeToDeletePtr;
+			NodeToDeletePtr = nullptr;
+
+		}
+
+		Size--;
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	void Clear()
+	{
+		while (Head)
+		{
+			Remove(0);
+		}
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	class FIterator
+	{
+	private:
+		FNode* Current;
+
+	public:
+		FIterator(FNode* NewCurrent) : Current(NewCurrent) {}
+
+		FIterator& operator++()
+		{
+			Current = Current->Next;
+			return *this;
+		}
+
+		T& operator*()
+		{
+			return Current->Data;
+		}
+
+		bool operator!=(const FIterator& Other)
+		{
+			return Current != Other.Current;
+		}
+	};
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	FIterator begin() const
+	{
+		return FIterator(Head);
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	FIterator end() const
+	{
+		return FIterator(nullptr);
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	template<typename Pred>
+	void ForEach(const Pred& Predicate)
+	{
+		for (T& Item : *this)
+		{
+			Predicate(Item);
+		}
```

Comment:
this is nice, I think I'm gonna borrow it `ctrl + c` `ctrl + v` xd

Reply:
MORE LIKE STEALIN', HUH